t3f/ops.py

```diff
@@ -1132,3 +1136,50 @@ def gather_nd(tt, indices):
     tt_elements = tf.matmul(tt_elements, core_slices)
   tt_elements = tf.reshape(tt_elements, tf.shape(indices)[:-1])
   return tt_elements
+
+def renormalize_cores(tt):
+    """Renormalizes tt cores to make them of the same Frobenius norm.
+
+    Does not change tt, but makes computations with cores more stable.
+
+    Args:
+      tt: `TensorTrain` or `TensorTrainBatch` object
+
+    Returns:
+      `TensorTrain` or `TensorTrainBatch` which represents the same
+      tensor as tt, but with all cores having equal norm. In the batch
+      case applies to each TT in `TensorTrainBatch`.
+
+    """
+    if isinstance(tt, TensorTrain):
+      new_cores = []
+      running_log_norm = 0
+      core_norms = []
+      for core in tt.tt_cores:
+        cur_core_norm = tf.sqrt(tf.reduce_sum(core ** 2) + 1e-8)
+        core_norms.append(cur_core_norm)
+        running_log_norm += tf.log(cur_core_norm)
+
+      running_log_norm = running_log_norm / tt.ndims()
+      fact = tf.exp(running_log_norm)
+      for i, core in enumerate(tt.tt_cores):
+        new_cores.append(core * fact / core_norms[i])
```

Comment:
Why do you need to divide by core_norms?

Reply:
To make each core of the norm "fact", which is equal to (||C_1|| ||C_2|| ... ||C_d||)^(1/d). So I divide each core by its norm, and multiply by "fact". This doesn't change overall tensor.

Reply:
Make sense, sorry for being stupid :)

Reply:
I don't see how to close this thread. If you don't either, just ignore it)